2003-08-05  Josef Zlomek  <zlomekj@suse.cz>

	* gcse.c (try_replace_reg): Fix updating of note.

--- gcc-3.3.1/gcc/gcse.c.gcse-fix	2003-07-16 14:37:54.000000000 +0200
+++ gcc-3.3.1/gcc/gcse.c	2003-08-05 17:48:43.000000000 +0200
@@ -3943,6 +3943,11 @@ try_replace_reg (from, to, insn)
 	validate_change (insn, &SET_SRC (set), src, 0);
     }
 
+  /* If there is already a NOTE, update the expression in it with our
+     replacement.  */
+  if (note != 0)
+    XEXP (note, 0) = simplify_replace_rtx (XEXP (note, 0), from, to);
+
   if (!success && set && reg_mentioned_p (from, SET_SRC (set)))
     {
       /* If above failed and this is a single set, try to simplify the source of
@@ -3963,11 +3968,6 @@ try_replace_reg (from, to, insn)
 	note = set_unique_reg_note (insn, REG_EQUAL, copy_rtx (src));
     }
 
-  /* If there is already a NOTE, update the expression in it with our
-     replacement.  */
-  else if (note != 0)
-    XEXP (note, 0) = simplify_replace_rtx (XEXP (note, 0), from, to);
-
   /* REG_EQUAL may get simplified into register.
      We don't allow that. Remove that note. This code ought
      not to hapen, because previous code ought to syntetize
